File: mini-yelp/mini-yelp.cc
/**
 * Definition of Location:
 * class Location {
 * public:
 *     double latitude, longitude;
 *     static Location create(double lati, double longi) {
 *         // This will create a new location object
 *     }
 * };
 * Definition of Restaurant
 * class Restaurant {
 * public:
 *     int id;
 *     string name;
 *     Location location;
 *     static Restaurant create(string &name, Location &location) {
 *         // This will create a new restaurant object,
 *         // and auto fill id
 *     }
 * };
 * Definition of Helper
 * class Helper {
 * public:
 *     static get_distance(Location &location1, Location &location2) {
 *         // return distance between location1 and location2.
 *     }
 * };
 * class GeoHash {
 * public:
 *     static string encode(Location &location) {
 *         // return convert location to a GeoHash string
 *     }
 *     static Location decode(string &hashcode) {
 *          // return convert a GeoHash string to location
 *     }
 * };
 */

typedef pair<double, string> pds;


class MiniYelp {
    public:
    MiniYelp() {
        // initialize your data structure here.
        codes.clear();
        rests.clear();
    }
    
    // @param name a string
    // @param location a Location
    // @return an integer, restaurant's id
    int addRestaurant(string name, Location &location) {
        // Write your code here
        Restaurant rest = Restaurant::create(name, location);
        string code = GeoHash::encode(location);
        stringstream ss;
        ss << rest.id;
        string suffix;
        ss >> suffix;
        code += "." + suffix;
        codes[rest.id] = code;
        rests[code] = rest;
        return rest.id;
    }
    
    // @param restaurant_id an integer
    void removeRestaurant(int restaurant_id) {
        // Write your code here
        string code = codes[restaurant_id];
        codes.erase(restaurant_id);
        rests.erase(code);
    }
    
    // @param location a Location
    // @param k an integer, distance smaller than k miles
    // @return a list of restaurant's name and sort by
    // distance from near to far.
    vector<string> neighbors(Location &location, double k) {
        // Write your code here
        string code = GeoHash::encode(location);
        
        int prefix = 12;
        for (int i = 0; i < 12; ++i) {
            if (k > errors[i]) {
                prefix = i;
                break;
            }
        }
        code = code.substr(0, prefix);
        priority_queue<pds, vector<pds>, greater<pds>> pq;
        map<string, Restaurant>::iterator it = rests.lower_bound(code);
        for (; it != rests.end(); ++it) {
            if (it->first.substr(0, prefix) != code) {
                break;
            }
            Restaurant trest = it->second;
            double tdist = Helper::get_distance(location, trest.location);
            if (tdist <= k) {
                pq.push(make_pair(tdist, trest.name));
            }
        }
        vector<string> ans;
        while (!pq.empty()) {
            ans.push_back(pq.top().second);
            pq.pop();
        }
        return ans;
    }
    
    private:
    map<int, string> codes;
    map<string, Restaurant> rests;
    double errors[12] = {2500, 630, 78, 20, 2.4, 0.61, 0.076, 0.01911, 0.00478, 0.0005971, 0.0001492, 0.0000186};
};

File: mini-yelp/mini-yelp.py
'''
    Definition of Location:
    class Location:
    # @param {double} latitude, longitude
    # @param {Location}
    @classmethod
    def create(cls, latitude, longitude):
    # This will create a new location object
    
    Definition of Restaurant:
    class Restaurant:
    # @param {str} name
    # @param {Location} location
    # @return {Restaurant}
    @classmethod
    def create(cls, name, location):
    # This will create a new restaurant object,
    # and auto fill id
    
    Definition of Helper
    class Helper:
    # @param {Location} location1, location2
    @classmethod
    def get_distance(cls, location1, location2):
    # return calculate the distance between two location
    
    Definition of GeoHash
    class GeoHash:
    # @param {Location} location
    # @return a string
    @classmethom
    def encode(cls, location):
    # return convert location to a geohash string
    
    # @param {str} hashcode
    # @return {Location}
    @classmethod
    def decode(cls, hashcode):
    # return convert a geohash string to location
    '''
from YelpHelper import Location, Restaurant, GeoHash, Helper
import bisect

class MiniYelp:
    
    def __init__(self):
        # initialize your data structure here.
        self.rests = {}
        self.codes = {}
        self.geos = []
    
    
    # @param {str} name
    # @param {Location} location
    # @return {int} restaurant's id
    def add_restaurant(self, name, location):
        # Write your code here
        rest = Restaurant.create(name, location)
        code = "%s.%s" % (GeoHash.encode(location), rest.id)
        self.codes[rest.id] = code
        self.rests[code] = rest
        bisect.insort(self.geos, code)
        return rest.id
    
    
    # @param {int} restaurant_id
    # @return nothing
    def remove_restaurant(self, restaurant_id):
        # Write your code here
        code = self.codes[restaurant_id]
        del self.rests[code]
        del self.codes[restaurant_id]
        self.geos.pop(bisect.bisect_left(self.geos, code))
    
    
    # @param {Location} location
    # @param {double} k, distance smaller than k miles
    # @return {str[]} a list of restaurant's name and sort by
    # distance from near to far.
    def neighbors(self, location, k):
        # Write your code here
        errors = [2500, 630, 78, 20, 2.4, 0.61, 0.076, 0.01911, 0.00478, 0.0005971, 0.0001492, 0.0000186]
        pre = 12
        for i in xrange(12):
            if k > errors[i]:
                pre = i
                break
        code = GeoHash.encode(location)[:pre]
        l = bisect.bisect_left(self.geos, code)
        r = bisect.bisect_right(self.geos, code + '{')
        ans = []
        for i in xrange(l, r):
            code = self.geos[i]
            rest = self.rests[code]
            dist = Helper.get_distance(location, rest.location)
            if dist <= k:
                ans.append((dist, rest.name))
        
        ans = sorted(ans, key=lambda x: x[0])
    return [x[1] for x in ans]
